[SPARK-559] Changes for strict mode CI tests

susanxhuynh opened the pull request:
Changes for the tests to run in strict mode
* Makefile change to spin up a strict cluster
* Setup service account, grant permissions
** Moved from old `setup_permissions.sh` script to sdk_security module
* Set a default "/spark" app name; granted permissions to drivers with that app name
* Marked test_jar() and test_marathon_group(), which runs test_jar(), as “expected fail”; failures tracked by SPARK-593

susanxhuynh commented:
cc @ArtRand @skonto

ArtRand commented on Makefile:
```diff
@@ -135,6 +135,7 @@ test-env:
 	source test-env/bin/activate
 	pip3 install -r tests/requirements.txt
 
+CF_TEMPLATE_URL ?= https://s3.amazonaws.com/downloads.mesosphere.io/dcos-enterprise/testing/master/cloudformation/ee.single-master.cloudformation.json
```
this is the strict CF template?

susanxhuynh replied:
This is permissive by default. The CI job in TeamCity (spark-build-PR-strict) has this set to a strict template.

ArtRand commented on tests/test_kafka.py:
```diff
@@ -140,7 +151,7 @@ def test_pipeline(kerberos_flag, stop_count, jar_uri, keytab_secret, jaas_uri=No
         "--conf", "spark.mesos.executor.secret.names={}".format(keytab_secret),
         "--conf", "spark.mesos.executor.secret.filenames=kafka-client.keytab",
-        "--conf", "spark.mesos.task.labels=DCOS_SPACE:/spark",
+        "--conf", "spark.mesos.task.labels=DCOS_SPACE:{}".format(utils.SPARK_APP_NAME),
```
This is done [in the CLI](https://github.com/mesosphere/spark-build/blob/master/cli/dcos-spark/submit_builder.go#L651) and iff you want to specify a different DCOS_SPACE, it should be done as a [DC/OS arg](https://github.com/mesosphere/spark-build/blob/master/cli/dcos-spark/main.go#L25) - I can make the PR to change this, however.

susanxhuynh replied:
OK, I'll let you change it in a separate PR.

ArtRand commented on tests/test_spark.py:
```diff
@@ -187,15 +187,15 @@ def test_s3():
                     args=args)
 
 
+@pytest.mark.xfail(utils.is_strict(), reason="Currently fails in strict mode")
```
We should change this test to something that can be exercised in `strict`, probably still good to test this as folding (marathon groups) is becoming the norm.

susanxhuynh replied:
Yeah, guess it doesn't matter much which test it is. I'll change it to SparkPi.

ArtRand commented on tests/utils.py:
```diff
@@ -2,18 +2,26 @@
 import shakedown
 import subprocess
 import urllib
+import urllib.parse
```
is there a reason this can't be done with the above import?

susanxhuynh replied:
If I only import `urllib`, the following line will fail:
`app_id_encoded = urllib.parse.quote(app_id, safe='')`

ArtRand commented on tests/utils.py:
```diff
@@ -175,11 +179,10 @@ def upload_file(file_path):
-        args += ["--conf", 'spark.mesos.driverEnv.MESOS_MODULES=file:///opt/mesosphere/etc/mesos-scheduler-modules/dcos_authenticatee_module.json']
-        args += ["--conf", 'spark.mesos.driverEnv.MESOS_AUTHENTICATEE=com_mesosphere_dcos_ClassicRPCAuthenticatee']
-        args += ["--conf", 'spark.mesos.principal=service-acct']
+        args += ["--conf", 'spark.mesos.containerizer=mesos']
```
is MESOS containerizer required for `strict`? (wasn't in the past)

susanxhuynh replied:
Let me double-check ... I don't remember why I added it.

ArtRand commented on tests/utils.py:
```diff
@@ -241,3 +244,55 @@ def teardown_spark():
+
+def spark_security_session():
+    '''
+    Spark strict mode setup is slightly different from dcos-commons, so can't use sdk_security::security_session.
```
explain why?

susanxhuynh replied:
I'll add more details about what's different: (1) the role is "*", (2) the driver permissions to execute tasks.

ArtRand commented on tests/utils.py:
```diff
@@ -241,3 +244,55 @@ def teardown_spark():
+    def grant_driver_permission(service_account_name, app_id):
+        dcosurl, headers = sdk_security.get_dcos_credentials()
+        app_id_encoded = urllib.parse.quote(app_id, safe='')
+        app_id_encoded = urllib.parse.quote(app_id_encoded, safe='')  # double-encoded (why?)
```
could you nest these so there isn't two assignments?

susanxhuynh replied:
Sure

ArtRand commented on tests/utils.py:
```diff
@@ -241,3 +244,55 @@ def teardown_spark():
     return s3.http_url(os.path.basename(os.environ["SCALA_TEST_JAR_PATH"]))
+
+
+def spark_security_session():
```
👍

susanxhuynh commented:
@ArtRand I remember why I added mesos containerizer: it was either that or needing to set SPARK_USER in docker containerizer. I've changed it to the latter, since that's our current default (docker).

susanxhuynh commented:
@ArtRand Addressed your comments. PTAL.

ArtRand commented:
LGTM Thanks @susanxhuynh
